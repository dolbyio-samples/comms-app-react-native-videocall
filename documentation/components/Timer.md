# Timer

The Timer component shows the elapsed time.

## Examples

### React

```javascript
return <Timer />;
```