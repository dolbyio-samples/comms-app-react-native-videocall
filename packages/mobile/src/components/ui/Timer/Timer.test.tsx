import React from 'react';
import { Text } from 'react-native';
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from 'react-test-renderer';

import { render } from '../../../utils/test-utils';

import Timer from './Timer';

const testID = 'testID';
const hourTestID = 'testID-hour';
const minTestID = 'testID-min';
const secTestID = 'testID-sec';

beforeEach(() => {
  // use fake timer
  jest.useFakeTimers();
});

afterEach(() => {
  // restore timer
  jest.useRealTimers();
});

describe('Timer component', () => {
  test('Passes TestID', () => {
    const { getByTestId } = render(<Timer testID={testID} />);

    const element = getByTestId(testID);
    expect(element).not.toBeNull();

    const textElmList = element.findAllByType(Text);
    expect(textElmList).toHaveLength(3);

    const hourElm = getByTestId(hourTestID);
    const minElm = getByTestId(minTestID);
    const secElm = getByTestId(secTestID);
    expect(hourElm).not.toBeNull();
    expect(minElm).not.toBeNull();
    expect(secElm).not.toBeNull();
  });
  test('display seconds value < 10', () => {
    const { getByTestId } = render(<Timer testID={testID} />);
    const hourElm = getByTestId(hourTestID);
    const minElm = getByTestId(minTestID);
    const secElm = getByTestId(secTestID);
    renderer.act(() => {
      // bump up timer value by 2 second
      jest.advanceTimersByTime(2000);
    });
    expect(hourElm.props.children).toBe(null);
    expect(minElm.props.children).toBe('00:');
    expect(secElm.props.children).toBe('02');
  });
  test('display seconds value > 10', () => {
    const { getByTestId } = render(<Timer testID={testID} />);
    const hourElm = getByTestId(hourTestID);
    const minElm = getByTestId(minTestID);
    const secElm = getByTestId(secTestID);
    renderer.act(() => {
      // bump up timer value by 59 second
      jest.advanceTimersByTime(59000);
    });
    expect(hourElm.props.children).toBe(null);
    expect(minElm.props.children).toBe('00:');
    expect(secElm.props.children).toBe(59);
  });
  test('display minutes value < 10', () => {
    const { getByTestId, rerender } = render(<Timer testID={testID} />);
    const hourElm = getByTestId(hourTestID);
    const minElm = getByTestId(minTestID);
    const secElm = getByTestId(secTestID);

    const totalTime = 2; // 02 minutes
    for (let i = 1; i <= totalTime; i++) {
      renderer.act(() => {
        jest.advanceTimersByTime(60000);
      });

      rerender(<Timer testID={testID} />);
    }
    expect(hourElm.props.children).toBe(null);
    expect(minElm.props.children).toBe('02:');
    expect(secElm.props.children).toBe('00');
  });
  test('display minutes value > 10', () => {
    const { getByTestId, rerender } = render(<Timer testID={testID} />);
    const hourElm = getByTestId(hourTestID);
    const minElm = getByTestId(minTestID);
    const secElm = getByTestId(secTestID);

    const totalTime = 11; // 11 minutes
    for (let i = 1; i <= totalTime; i++) {
      // each rendering increments the timer value by 1 minute
      renderer.act(() => {
        jest.advanceTimersByTime(60000);
      });

      rerender(<Timer testID={testID} />);
    }
    expect(hourElm.props.children).toBe(null);
    expect(minElm.props.children).toBe('11:');
    expect(secElm.props.children).toBe('00');
  });
  test('display hours value < 10', () => {
    const { getByTestId, rerender } = render(<Timer testID={testID} />);
    const hourElm = getByTestId(hourTestID);
    const minElm = getByTestId(minTestID);
    const secElm = getByTestId(secTestID);

    const totalTime = 60; // 1 hours
    for (let i = 1; i <= totalTime; i++) {
      // each rendering increments the timer value by 60 seconds
      renderer.act(() => {
        jest.advanceTimersByTime(60000);
      });

      rerender(<Timer testID={testID} />);
    }
    expect(hourElm.props.children).toBe('01:');
    expect(minElm.props.children).toBe('00:');
    expect(secElm.props.children).toBe('00');
  });
  test('display hours value > 10', () => {
    const { getByTestId, rerender } = render(<Timer testID={testID} />);
    const hourElm = getByTestId(hourTestID);
    const minElm = getByTestId(minTestID);
    const secElm = getByTestId(secTestID);
    const totalTime = 60 * 11; // 11 hours
    for (let i = 1; i <= totalTime; i++) {
      // each rendering increments the timer value by 60 seconds
      renderer.act(() => {
        jest.advanceTimersByTime(60000);
      });

      rerender(<Timer testID={testID} />);
    }
    expect(hourElm.props.children).toBe('11:');
    expect(minElm.props.children).toBe('00:');
    expect(secElm.props.children).toBe('00');
  });

  test('display hours:minutes:seconds value', () => {
    const { getByTestId, rerender } = render(<Timer testID={testID} />);
    const hourElm = getByTestId(hourTestID);
    const minElm = getByTestId(minTestID);
    const secElm = getByTestId(secTestID);

    let totalTime = 60; // 1 hours
    for (let i = 1; i <= totalTime; i++) {
      // each rendering increments the timer value by 60 seconds
      renderer.act(() => {
        jest.advanceTimersByTime(60000);
      });

      rerender(<Timer testID={testID} />);
    }

    // change minutes value
    totalTime = 2; // 02 minutes
    for (let i = 1; i <= totalTime; i++) {
      // each rendering increments the timer value by 1 minute
      renderer.act(() => {
        jest.advanceTimersByTime(60000);
      });

      rerender(<Timer testID={testID} />);
    }

    // change second value
    renderer.act(() => {
      jest.advanceTimersByTime(9000);
    });
    rerender(<Timer testID={testID} />);

    expect(hourElm.props.children).toBe('01:');
    expect(minElm.props.children).toBe('02:');
    expect(secElm.props.children).toBe('09');
  });
});

describe('Timer component snapshot', () => {
  test('should match default', () => {
    const tree = renderer.create(<Timer testID={testID} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
