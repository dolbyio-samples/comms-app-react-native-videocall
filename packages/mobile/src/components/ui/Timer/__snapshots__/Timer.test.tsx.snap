// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer component snapshot should match default 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
        "flexDirection": "row",
      },
    ]
  }
  testID="testID"
>
  <Text
    style={
      Array [
        Object {
          "letterSpacing": 1.22,
        },
        Object {
          "fontSize": 12,
          "fontWeight": "600",
          "letterSpacing": 0.3,
          "lineHeight": 16,
        },
        false,
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Roboto",
          "textAlign": "left",
        },
      ]
    }
    testID="testID-hour"
  />
  <Text
    style={
      Array [
        Object {
          "letterSpacing": 1.22,
        },
        Object {
          "fontSize": 12,
          "fontWeight": "600",
          "letterSpacing": 0.3,
          "lineHeight": 16,
        },
        false,
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Roboto",
          "textAlign": "left",
        },
      ]
    }
    testID="testID-min"
  >
    00:
  </Text>
  <Text
    style={
      Array [
        Object {
          "letterSpacing": 1.22,
        },
        Object {
          "fontSize": 12,
          "fontWeight": "600",
          "letterSpacing": 0.3,
          "lineHeight": 16,
        },
        false,
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Roboto",
          "textAlign": "left",
        },
      ]
    }
    testID="testID-sec"
  >
    00
  </Text>
</View>
`;
